[WIP] Trying to allow plotting SEEG activity in volumetric source space

adam2392 opened the pull request:
## Reference issue
Fixes: #8364 
Closes #8388


#### What does this implement/fix?
1. Adds a `plot_seeg.py` tutorial for working w/ SEEG data analogous to `plot_ecog.py`.

2. Attempts to refactor `stc_near_sensors` to show volumetric SEEG activity with either `stc.plot` or `stc.plot_3d`


#### Additional information
Opening up PR to start iterating.

Ideally want to allow SEEG activity to be visualized as such: https://mne.tools/dev/auto_tutorials/source-modeling/plot_beamformer_lcmv.html#volumetric-rendering-3d
Assumes user has:
- electrode locations
- raw SEEG activity

Apologies if the code is nowhere where it should be. Never worked with the source API on mne, but now am eager to learn it if it'll help produce these nice viz of SEEG.

larsoner commented on mne/source_estimate.py:
```diff
@@ -3196,6 +3196,12 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
     %(subjects_dir)s
+    src : instance of SourceSpaces | instance of SourceMorph
+        The source space. Can also be a SourceMorph to morph the STC to
+        a new subject (see Examples).
```
I would not allow `SourceMorph` here, just `src`. People can morph ahead of time if they want.

larsoner commented on mne/source_estimate.py:
```diff
@@ -3231,9 +3237,18 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
+    else:
+        evoked = evoked.copy().pick_types(meg=False, seeg=True)
+    # get channel positions and coord_frame transformation
+    # XXX: how to allow MNI?
+    pos = evoked._get_channel_positions()
```
The channel positions should always be in the head coordinate frame. Then we use `trans` to get them to the MRI coordinate frame.

People with coordinates in MNI coords should just use `fsaverage` as their subject since fsaverage is in MNI space (i.e., MNI==fsaverage's MRI coordinate frame)

larsoner commented on mne/source_estimate.py:
```diff
@@ -3243,6 +3258,8 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
     rrs = np.concatenate(rrs)
     rrs /= 1000.
+
+    # convert head positions -> native coord_frame (e.g. mni_tal, or mri)
```
No, always `MRI`

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,137 @@
+.. _tut_working_with_seeg:
+
+======================
+Working with SEEG data
```
```suggestion
Working with sEEG data
```

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,137 @@
+
+MNE supports working with more than just MEG and EEG data. Here we show some
+of the functions that can be used to facilitate working with
+stereoelectroencephalography (SEEG) data.
```
```suggestion
stereoelectroencephalography (sEEG) data.
```

larsoner commented on mne/source_estimate.py:
```diff
@@ -3176,7 +3176,7 @@ def extract_label_time_course(stcs, labels, src, mode='auto',
 def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
-                     project=True, subjects_dir=None, verbose=None):
+                     project=True, subjects_dir=None, src=None, verbose=None):
     """Create a STC from ECoG sensor data.
```
ECoG and sEEG

larsoner commented on mne/source_estimate.py:
```diff
@@ -3231,9 +3237,18 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
-    pos = np.array([ch['loc'][:3] for ch in evoked.info['chs']])
+
+    # create a copy of Evoked using either ecog or seeg
+    if src is None:
+        evoked = evoked.copy().pick_types(meg=False, ecog=True)
+    else:
+        evoked = evoked.copy().pick_types(meg=False, seeg=True)
```
```suggestion
    evoked = evoked.copy().pick_types(ecog=True, seeg=True)
```

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,137 @@
+# paths to mne datasets - sample SEEG and FreeSurfer subject
+misc_path = mne.datasets.misc.data_path()
+sample_path = mne.datasets.sample.data_path()
+subject = 'sample'
+subjects_dir = sample_path + '/subjects'
```
For this example, do not use `sample`. Use `fsaverage`. That way the ECoG example is "ECoG data in a subject's space projected to a surface" and the sEEG example is "sEEG data in MNI space projected to a volume". And people who have ECoG/sEEG x subject/MNI x surf/vol can pick the parts of the tutorials they need. And the narrative docs of the two examples should begin by stating this very clearly, e.g., this should have something like:
```
This example shows how to use:

- sEEG data
- channel locations in MNI space
- projection into a volume

For an example that involves ECoG data, channel locations in a subject-specific MRI, or projection into a surface, see :ref:`tut-working-with-ecog`.
```
And then `plot_ecog.py` can be updated with a similar list+ref.

larsoner reviewed (commented):
Looks like a good start!

adam2392 commented:
I keep on getting this error, and I'm not sure how to trace this down. It seems that there is no data getting attached to `stc.data`? 

```
Traceback (most recent call last):
  File "/Users/adam2392/Documents/mne-python/tutorials/misc/plot_seeg.py", line 142, in <module>
    brain = stc.plot(src=src, mode='stat_map', initial_time=0.68,
  File "/Users/adam2392/Documents/mne-python/mne/source_estimate.py", line 1943, in plot
    return plot_volume_source_estimates(
  File "<decorator-gen-140>", line 24, in plot_volume_source_estimates
  File "/Users/adam2392/Documents/mne-python/mne/viz/_3d.py", line 2104, in plot_volume_source_estimates
    img = stc.as_volume(src, mri_resolution=False)
  File "/Users/adam2392/Documents/mne-python/mne/source_estimate.py", line 2106, in as_volume
    return _interpolate_data(data, src, mri_resolution=mri_resolution,
  File "/Users/adam2392/Documents/mne-python/mne/morph.py", line 823, in _interpolate_data
    vols[this_inuse] = stc.data[stc_slice]
ValueError: shape mismatch: value array of shape (0,2001) could not be broadcast to indexing result of shape (24303,2001)
```

larsoner commented on mne/source_estimate.py:
```diff
@@ -3231,19 +3238,32 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
+    # in the Source space we will use the evoked data
+    pos = evoked._get_channel_positions()
+
+    # coord_frame transformation from native mne "head" to desired coord_frame
+    # coord_frame (e.g. 'mni_tal', 'mri', 'fs_tal', etc.)
+    trans, _ = _get_trans(trans, 'head', coord_frame, allow_none=True)
```
```suggestion
    # coord_frame transformation from native mne "head" to MRI coord_frame
    trans, _ = _get_trans(trans, 'head', coord_frame, allow_none=True)
```

larsoner commented on mne/source_estimate.py:
```diff
@@ -3231,19 +3238,32 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
+    # coord_frame (e.g. 'mni_tal', 'mri', 'fs_tal', etc.)
+    trans, _ = _get_trans(trans, 'head', coord_frame, allow_none=True)
+
+    # convert head positions -> native coord_frame MRI
+    pos = apply_trans(trans, pos)
```
```suggestion
    # convert head positions -> coord_frame MRI
    pos = apply_trans(trans, pos)
```

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,149 @@
+- channel locations in MNI space
+- projection into a volume
+
+For an example that involves ECoG data, channel locations in a subject-specific MRI, or projection into a surface, see :ref:`tut-working-with-ecog`.
```
PEP8
```suggestion
For an example that involves ECoG data, channel locations in a
subject-specific MRI, or projection into a surface, see
:ref:`tut-working-with-ecog`.
```

larsoner commented on tutorials/misc/plot_ecog.py:
```diff
@@ -8,6 +8,14 @@
+- channel locations in subject's MRI space
+- projection onto a surface
+
+For an example that involves sEEG data, channel locations in a template-specific MRI, or projection into a volume, see :ref:`tut-working-with-seeg`.
```
```suggestion
For an example that involves sEEG data, channel locations in
MNI space, or projection into a volume, see :ref:`tut-working-with-seeg`.
```

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,149 @@
+# set channel types to sEEG (instead of EEG)
+raw.set_channel_types({ch_name: 'seeg' for ch_name in raw.ch_names})
+
+###############################################################################
+# We can then plot the locations of our electrodes on our subject's brain.
+# We'll use :func:`~mne.viz.snapshot_brain_montage` to save the plot as image
+# data (along with xy positions of each electrode in the image), so that later
+# we can plot frequency band power on top of it.
+#
+# .. note:: These are not real electrodes for this subject, so they
+#           do not align to the cortical surface perfectly.
+
+# fig = plot_alignment(raw.info, subject=subject, subjects_dir=subjects_dir,
+#                      surfaces=['pial'], trans=trans, coord_frame=coord_frame)
+# mne.viz.set_3d_view(fig, 200, 70, focalpoint=[0, -0.005, 0.03])
+#
+# xy, im = snapshot_brain_montage(fig, montage)
```
Delete, and mention this difference at the top

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,149 @@
+# xy, im = snapshot_brain_montage(fig, montage)
+
+
+###############################################################################
+# Next, we'll compute the signal power in the gamma (30-90 Hz) and alpha
+# (8-12 Hz) bands.
+gamma_power_t = raw.copy().filter(30, 90).apply_hilbert(
+    envelope=True).get_data()
+alpha_power_t = raw.copy().filter(8, 12).apply_hilbert(
+    envelope=True).get_data()
+gamma_power = gamma_power_t.mean(axis=-1)
+alpha_power = alpha_power_t.mean(axis=-1)
+
+# choose a colormap range wide enough for both frequency bands
+_gamma_alpha_power = np.concatenate((gamma_power, alpha_power)).flatten()
+vmin, vmax = np.percentile(_gamma_alpha_power, [10, 90])
```
Maybe we should just plot raw traces in this example?

adam2392 replied:
Okay took this part out, and just added a `raw.plot()` to show that plotting SEEG data raw.

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,149 @@
+# sphinx_gallery_thumbnail_number = 4
+
+# setup a volume-based source space here
+vol_src = mne.setup_volume_source_space(subject, subjects_dir=subjects_dir,
+                                        pos=5., mri='aseg.mgz')
```
you should not need to do this for fsaverage. Instead do:
```suggestion
# get standard fsaverage volume source space
fetch_fsaverage(subjects_dir=subjects_dir)  # downloads it if necessary
fname_src = op.join(subjects_dir, 'fsaverage', 'bem',
                   'fsaverage-vol-5-src.fif')
vol_src = mne.read_source_spaces(fname_src)
```

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,149 @@
+
+###############################################################################
+# Now we make a :class:`mne.channels.DigMontage` stating that the sEEG
+# contacts are in the FreeSurfer surface RAS (i.e., MRI) coordinate system.
```
```suggestion
# contacts are in the FreeSurfer surface RAS (i.e., MRI) coordinate system
# for the given subject. Keep in mind that ``fsaverage`` is special in that
# it is already in MNI space.
```

larsoner commented on mne/source_estimate.py:
```diff
@@ -3196,6 +3197,12 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
         If True, project the electrodes to the nearest ``'pial`` surface
         vertex before computing distances.
```
```suggestion
        vertex before computing distances. Only used when doing a
        surface projection.
```

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,149 @@
+
+evoked = mne.EvokedArray(gamma_power_t, raw.info)
+stc = mne.stc_near_sensors(evoked, trans, subject, subjects_dir=subjects_dir,
+                           src=vol_src, project=False, mode='nearest')
```
```suggestion
                           src=vol_src, mode='nearest')
```

larsoner reviewed (commented):
Would it help if I pulled your changes and fixed the problem you're having with not having any data?

larsoner commented:
> Do you think this is a problem with the seeg coordinates, or possibly something with the stc_near_sensors function?

I suspect that your `evoked.data.shape[1] == 0` so your `stc.data.shape[1] == 0` and the volume morphing and stc plotting functions don't handle this case properly.

adam2392 commented:
Okay a few updates:

1. it turned out that channel coords for the `misc-data` was in 'mm', so needed to divide by 1000 before setting to montage. 
2. In addition, if channels have `nan` coordinates, we need to remove to make sure `pdist(pos)` doesn't return nan. 
3. I added a bunch of inline comments to help navigate a bit better.
4. I added a bunch of print statements to debug, but I can remove them once I figure out why the following error is occuring.

```
Traceback (most recent call last):
  File "/Users/adam2392/Documents/mne-python/tutorials/misc/plot_seeg.py", line 129, in <module>
    brain = stc.plot(src=vol_src, mode='stat_map', initial_time=0.68,
  File "/Users/adam2392/Documents/mne-python/mne/source_estimate.py", line 1943, in plot
    return plot_volume_source_estimates(
  File "<decorator-gen-140>", line 24, in plot_volume_source_estimates
  File "/Users/adam2392/Documents/mne-python/mne/viz/_3d.py", line 2104, in plot_volume_source_estimates
    img = stc.as_volume(src, mri_resolution=False)
  File "/Users/adam2392/Documents/mne-python/mne/source_estimate.py", line 2106, in as_volume
    return _interpolate_data(data, src, mri_resolution=mri_resolution,
  File "/Users/adam2392/Documents/mne-python/mne/morph.py", line 823, in _interpolate_data
    vols[this_inuse] = stc.data[stc_slice]
ValueError: shape mismatch: value array of shape (1638,2001) could not be broadcast to indexing result of shape (14629,2001)
```

It's trying to get 14629 vertices, which I'm guessing is coming from the `vol_src` for `fsaverage`, but I don't see anyway of limiting the amount of vertices there. In `stc_near_sensors`, we only grab 1638 vertices within our "epsilon-ball" parametrized by `distance=0.01`? Any idea how I should proceed? I don't think it seems to be a bug anymore on my end...

adam2392 commented:
Awesome @larsoner I just saw the recent changes and it works now. Thanks for the help! It looks very cool.

Regarding what is then plotted later, I see that it is kind of sparse but focused in the left frontal lobe region... I wonder if the example should talk a bit about it and what's the main orange region vs the blips that might appear? Even for me working on this example, I'm not sure what is being visualized per say without trying to go thru the docs. I imagine for a newcomer that would be useful too and then they would understand how to visualize a different set of contacts, or regions.

larsoner commented:
> In stc_near_sensors, we only grab 1638 vertices within our "epsilon-ball" parametrized by distance=0.01? Any idea how I should proceed? I don't think it seems to be a bug anymore on my end...

Basically if `src` is passed, the output should always have `stc.shape == (n_vertices, n_times)` where `n_vertices = sum(len(s['vertno']) for s in src)`, i.e., all source points should be there. It's just a lot of them will be zero.

I took a stab at simplifying / unifying the code. Feel free to take a look and see if it makes sense. This is the rendered tutorial now:

https://22913-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/misc/plot_seeg.html

Locally the `plot_alignment` plot would not look correct because your branch did not have #8393, so I also did a `merge` for your branch with `upstream/master`. Locally if you just do `git pull` you should get all my changes + this merge and be able to continue making changes.

larsoner commented:
> I see that it is kind of sparse but focused in the left frontal lobe region...

Take a look at the `plot_alignment` plot locally. This is the CircleCI version:

![sphx_glr_plot_seeg_001](https://user-images.githubusercontent.com/2365790/97014507-67372e00-1518-11eb-9950-cddbcee032e7.png)

If you look at it locally you can see that actually a lot of points are outside the scalp surface. Not sure why this would be the case if your points are in MNI space...

adam2392 commented:
Just took a look and those coordinates are in the wrong space. I updated the misc-data repo:

https://github.com/mne-tools/mne-misc-data/pull/4

adam2392 commented:
New ones rendered locally look accurate according to the test subject I was using to update misc-data.

<img width="825" alt="Screen Shot 2020-10-23 at 10 43 57 AM" src="https://user-images.githubusercontent.com/3460267/97018146-cac35a80-151c-11eb-91fd-a628afa49a9c.png">
<img width="678" alt="Screen Shot 2020-10-23 at 10 44 25 AM" src="https://user-images.githubusercontent.com/3460267/97018149-cc8d1e00-151c-11eb-8134-66381e9ab7e2.png">
<img width="638" alt="Screen Shot 2020-10-23 at 10 44 35 AM" src="https://user-images.githubusercontent.com/3460267/97018156-ce56e180-151c-11eb-92dd-9c65d8bb21e1.png">

BarryLiu-97 commented:
If you guys need sEEG data and its's electrodes MNI coordinates, I can offer one sample in .edf but it is too big. The least one is about 300MB.

BarryLiu-97 commented:
I can cut it then export it in .set format or you know, .fif format.

BarryLiu-97 commented:
![Uploading image.png…]()

adam2392 commented:
> If you guys need sEEG data and its's electrodes MNI coordinates, I can offer one sample in .edf but it is too big. The smallest one is about 300MB.

Any chance you can crop it to like 3-5 second so it's essentially a few MB at most? and then make a PR into the `seeg/` folder here: https://github.com/mne-tools/mne-misc-data?

We can have multiple SEEG example files then perhaps? @larsoner would that be valuable?

larsoner commented:
> We can have multiple SEEG example files then perhaps? @larsoner would that be valuable?

I'm not sure what the added value would be. Let's finish this example first, then see if there is something missing that we cannot cover adequately with this dataset and the ECoG dataset.

BarryLiu-97 commented:
> Any chance you can crop it to like 3-5 second so it's essentially a few MB at most? and then make a PR into the `seeg/` folder here: https://github.com/mne-tools/mne-misc-data?
Yeah, of course. Although I don't know how to make a PR into the `seeg/` folder, I'll give it a try. While as to sEEG data, some of its channels may not be valuable in analysis, for the reference electrode and the DC channel. And in my sample data, some of these DC channels are used to mark data. You want the original data directly from the machine, or I delete some useless channels and then mark the data.

BarryLiu-97 commented:
> > Any chance you can crop it to like 3-5 second so it's essentially a few MB at most? and then make a PR into the seeg/ folder here: https://github.com/mne-tools/mne-misc-data?
Yeah, of course. Although I don't know how to make a PR into the seeg/ folder, I'll give it a try. While as to sEEG data, some of its channels may not be valuable in analysis, for the reference electrode and the DC channel. And in my sample data, some of these DC channels are used to mark data. You want the original data directly from the machine, or I delete some useless channels and then mark the data.

BarryLiu-97 commented:
> Any chance you can crop it to like 3-5 second so it's essentially a few MB at most? and then make a PR into the `seeg/` folder here: https://github.com/mne-tools/mne-misc-data?

Yeah, of course. But as to sEEG data, some of its' electrodes are useless in analysis, like ref electrode, or EKG channels, DC channels. You want the data directly from the machine or I delete the useless channels and set the marker then give you?

BarryLiu-97 commented:
OK, if you guys need it, just tell me.

adam2392 commented:
Okay the new example looks good after applying the `talairach.xfm`, which I've stuck in the PR in `mne-misc-data`. Once that gets merged in and we update the md5sum, then the docs can be rendered here.

larsoner commented:
@adam2392 now you need to update `mne/datasets/utils.py` to use version 0.7 and hash 2b2f2fec9d1197ed459117db1c6341ee for MISC

larsoner commented on mne/tests/test_source_estimate.py:
```diff
@@ -1538,16 +1538,3 @@ def test_stc_near_sensors(tmpdir):
-                                   distance=0.033)
-    assert isinstance(stc_vol, VolSourceEstimate)
-    log = log.getvalue()
-    assert '4157 volume vertices' in log
```
Looks like you killed the volumetric test I added?

adam2392 replied:
oops sorry merging gone wrong.

larsoner commented on mne/tests/test_source_estimate.py:
```diff
@@ -1538,3 +1538,16 @@ def test_stc_near_sensors(tmpdir):
+    assert '4157 volume vertices' in log
+
+
+run_tests_if_main()
```
This part I actually removed

larsoner replied:
(This line and the one above it I mean)

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,162 @@
+# plot Nutmeg style
+stc.plot(vol_src, subject=subject, subjects_dir=subjects_dir, clim=clim)
+
+# plot 3D source (brain region) visualization
```
Create a separate section in the rendered doc:
```suggestion
###############################################################################
# Plot 3D source (brain region) visualization:
```

larsoner commented on tutorials/misc/plot_seeg.py:
```diff
@@ -0,0 +1,162 @@
+stc = abs(stc)  # just look at magnitude
+clim = dict(kind='value', lims=np.percentile(abs(evoked.data), [10, 50, 75]))
+
+# XXX: what is the difference between plot and plot_3d? They look visually
+# pretty similar.
+# plot Nutmeg style
```
```suggestion
# plot Nutmeg style
```

larsoner reviewed (commented):
Anything else left to do from your end @adam2392 or are you happy with this now?

larsoner commented:
Look like the `fetch_fsaverage` needs to be called at the top, see the error:

https://22951-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/misc/plot_seeg.html

It's not great that we got `CircleCI received exit code 0` from that build step...

pushed a commit to fix these

adam2392 commented:
> Anything else left to do from your end @adam2392 or are you happy with this now?

No this looks really good to me so far and honestly really useful for someone in the iEEG world to use mne-python in conjunction w/ freesurfer + neuroimaging viz.

I updated the docstring of `stc_near_sensors`. lmk wyt?

larsoner commented on mne/source_estimate.py:
```diff
@@ -3208,7 +3208,11 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
+        will have the same number of vertices as the passed in
+        SourceSpace object. If using surface, then ``stc`` will
+        have the number of vertices on the ``pial`` surface that
+        get activated.
```
This is actually not quite accurate.

If src is None, a surface source estimate will be produced, and the number of vertices will equal the number of pial-surface vertices that were close enough to the sensors to take on a non-zero volue.

If src is not None, a surface, volume, or mixed source estimate will be produced (depending on the kind of source space passed) and the vertices will match those of ``src`` (i.e., there may be me many all-zero values in ``stc.data``).

adam2392 replied:
Okay that makes more sense. I'll add that in.

adam2392 commented:
```
/home/circleci/project/mne/source_estimate.py:docstring of mne.stc_near_sensors:55: WARNING: Unexpected indentation.
```

I think this is occurring in the addition I made under `Notes`. I think(?) I fixed it?

larsoner commented on mne/source_estimate.py:
```diff
@@ -3194,21 +3195,33 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
-
+    For surface projections, this function projects the ECoG sensors to
+    the pial surface (if ``project``), then the activation at each pial
+    surface vertex is given by the mode:
```
You probably need to restore the newline here for rst to render properly. The error you saw was because of an extra space indentation that I pushed a fix for

larsoner commented on mne/source_estimate.py:
```diff
@@ -3222,7 +3222,6 @@ def stc_near_sensors(evoked, trans, subject, distance=0.01, mode='sum',
     For surface projections, this function projects the ECoG sensors to
     the pial surface (if ``project``), then the activation at each pial
     surface vertex is given by the mode:
-
```
In other words, this is probably going to break things

adam2392 replied:
Fixed.

adam2392 commented:
Another small thing: do you happen to know how one would do this if you don't have BEM? My understanding is you only have BEM from flash MRI data. I've never worked w/ BEM and it never gets output from my freesurfer `recon-all`.

```
# get standard fsaverage volume (5mm grid) source space
fname_src = op.join(subjects_dir, 'fsaverage', 'bem',
                    'fsaverage-vol-5-src.fif')
vol_src = mne.read_source_spaces(fname_src)
```

E.g. what if the above someone wanted to just use the parcellated MRI? Is that possible?

larsoner reviewed (approved):
LGTM +1 for merge

Let's see if @drammock and/or @agramfort are happy. Here is the rendered example:

https://22968-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/misc/plot_seeg.html

Just pushed a tiny commit to make the thumbnail the rendered movie.

larsoner commented:
> Another small thing: do you happen to know how one would do this if you don't have BEM? My understanding is you only have BEM from flash MRI data. I've never worked w/ BEM and it never gets output from my freesurfer recon-all.

This is another thing that we should make clearer somewhere in the docs. You only need a BEM if you want to do some forward modeling.

What you really probably want is to constrain your volumetric sources to be within the brain volume. One way to do this is by passing a BEM, but really this just uses the `inner_skull` surface. Instead of doing this, you can use the `brain` surface, which you can extract from the `brainmask.mgz` file using a few FreeSurfer commands -- this is done in the process of `mne watershed_bem`, so this is a convenient way to get this surface. This does not require FLASH scans -- it uses only the T1.mgz image.

Alternatively, you can tell it to use `aseg.mgz` or `aseg+aparc.mgz` with `single_volume=True`, which will only include volumetric sources that are in the brain volume / `brainmask.mgz`. Maybe open a separate issue about how to create and work with volumetric source spaces so that we can discuss this stuff there?

BarryLiu-97 commented:
Hi, is there any way that we can set different parts of depth electrodes into different colors, for example, which may be useful, if the contacts of this electrode is out of the skull, it's green, otherwise yellow.

larsoner commented:
> Hi, is there any way that we can set different parts of depth electrodes into different colors, for example, which may be useful, if the contacts of this electrode is out of the skull, it's green, otherwise yellow.

This is really an issue for / part of #8382

larsoner commented:
Thanks for working through this @adam2392 !

BarryLiu-97 commented:
Hi, recently I've been thinking that showing some info about the SEEG electrodes may be very useful when you move your mouse to the electrode in the figure produced by plot_alignment, for example, the name, the MNI coordinates, and which brain area the electrode is in. I have figured out how
 to calculate the brain area of each electrode, but I don't know how to add it to this.

BarryLiu-97 commented:
![image](https://user-images.githubusercontent.com/65115609/100820733-d2660f80-3489-11eb-9826-e4ae727ef8e0.png)

BarryLiu-97 commented:
This is the info of the electrodes I calculated using python.
![image](https://user-images.githubusercontent.com/65115609/100820809-03464480-348a-11eb-8c06-3e371c645622.png)

drammock commented:
@BarryLiu-97 please open a new issue for new feature requests. This PR is already merged and closed.

BarryLiu-97 commented:
> please open a new issue for new feature requests. This PR is already merged and closed.

Got it!
